commit a4ad1d659fbcd32d4beb25aa36b16f50db3806ad (refs/changes/52/2652/1)
Author: Kody A Kantor <kody.kantor@gmail.com>
Date:   2017-09-25T21:39:31+00:00 (2 years ago)
    
    joyent/node-fast#9 node-fast could track basic request metrics

diff --git a/CHANGES.md b/CHANGES.md
index def0764..07187cd 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,10 @@
 
 No changes.
 
+## v2.3.4
+
+* #9 node-fast could track basic request metrics
+
 ## v2.2.4
 
 * #7 fsr_context isn't always unpiped from fsr_encoder, which causes memory leaks
diff --git a/README.md b/README.md
index 7c4ce35..ba3f07f 100644
--- a/README.md
+++ b/README.md
@@ -97,6 +97,25 @@ The client and server only provide functions; you have to hook this up to a kang
 server.  The built-in demo server ("fastserve") and benchmarking tool
 ("fastbench") demonstrate how to do that.
 
+The Kang and metric collection (see below) functionality are served through the
+same [restify](https://github.com/restify/restify) server in "fastserve," but
+that is not required. Optionally, Kang can be served using the `knStartServer`
+function provided by the Kang library.
+
+### Metric Collection
+
+The Fast server optionally collects some basic request metrics using
+[node-artedi](https://github.com/joyent/node-artedi).
+
+The metrics collected by the fast server are:
+
+* total request count
+* latency of requests, implemented as a histogram
+
+You can pass an artedi-style metric collector into the Fast server constructor
+to enable metric collection. An example of how to do this is included in the
+built-in demo server ("fastserve").
+
 ### DTrace probes
 
 The Fast client and server provide DTrace probes and scripts in the "bin"
diff --git a/bin/fastserve b/bin/fastserve
index bd4e85c..ff39d4e 100755
--- a/bin/fastserve
+++ b/bin/fastserve
@@ -7,7 +7,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -15,6 +15,7 @@
  */
 
 var mod_assertplus = require('assert-plus');
+var mod_artedi = require('artedi');
 var mod_bunyan = require('bunyan');
 var mod_cmdutil = require('cmdutil');
 var mod_fastdemo = require('../lib/demo_server');
@@ -22,8 +23,9 @@ var mod_fastserver = require('../lib/fast_server');
 var mod_kang = require('kang');
 var mod_net = require('net');
 var mod_os = require('os');
+var mod_restify = require('restify');
 
-var KANG_PORT = 16519;
+var MONITOR_PORT = 16519;
 
 var usageMessage = [
     'Start a Fast protocol server supporting a few basic RPC calls.  PORT is ',
@@ -52,18 +54,24 @@ function main()
 
 	fastDemoServer({
 	    'fastPort': port,
-	    'kangPort': KANG_PORT
+	    'monitorPort': MONITOR_PORT
 	});
 }
 
 function fastDemoServer(args)
 {
-	var fastPort, log, sock, fastserver;
-	var kangPort, kangserver;
+	var fastPort, log, sock, collector, fastserver;
+	var monitorPort, kangserver;
 
 	mod_assertplus.object(args, 'args');
 	mod_assertplus.number(args.fastPort, 'args.fastPort');
-	mod_assertplus.number(args.kangPort, 'args.kangPort');
+	mod_assertplus.number(args.monitorPort, 'args.monitorPort');
+
+	collector = mod_artedi.createCollector({
+	    'labels': {
+		'component': 'fastserve'
+	    }
+	});
 
 	log = new mod_bunyan({
 	    'name': 'fastserve',
@@ -74,7 +82,8 @@ function fastDemoServer(args)
 	sock = mod_net.createServer({ 'allowHalfOpen': true });
 	fastserver = new mod_fastserver.FastServer({
 	    'log': log,
-	    'server': sock
+	    'collector': collector,
+	    'server': sock,
 	});
 
 	mod_fastdemo.demoRpcs().forEach(function (r) {
@@ -82,15 +91,14 @@ function fastDemoServer(args)
 	});
 
 	fastPort = args.fastPort;
-	kangPort = args.kangPort;
+	monitorPort = args.monitorPort;
 	sock.listen(fastPort, function () {
 		var nsigs = 0;
 
 		log.info({ 'fastPort': fastPort },
 		    'listening for fast requests');
 
-		mod_kang.knStartServer({
-		    'port': kangPort,
+		var kangOpts = {
 		    'uri_base': '/kang',
 		    'service_name': 'fastserve',
 		    'version': '1.0.0',
@@ -99,20 +107,34 @@ function fastDemoServer(args)
 		    'list_objects': fastserver.kangListObjects.bind(fastserver),
 		    'get': fastserver.kangGetObject.bind(fastserver),
 		    'stats': fastserver.kangStats.bind(fastserver)
-		}, function (err, server) {
-			mod_assertplus.ok(!err);
-			kangserver = server;
-			log.info({ 'kangPort': kangPort },
-			    'listening for kang requests');
-			if (nsigs > 0) {
-				kangserver.close();
-			}
+		};
+
+		var monitor_server = mod_restify.createServer({
+		    name: 'monitor'
+		});
+		monitor_server.get('/metrics', function (req, res, next) {
+		    collector.collect(mod_artedi.FMT_PROM,
+			function (err, metrics) {
+				if (err) {
+					next(err);
+					return;
+				}
+				res.setHeader('Content-Type',
+				    'text/plain; version=0.0.4');
+				res.send(metrics);
+				next();
+			});
+		});
+		monitor_server.get('/kang/.*',
+		    mod_kang.knRestifyHandler(kangOpts));
+		monitor_server.listen(monitorPort, '0.0.0.0', function () {
+			log.info({ 'monitorPort': monitorPort },
+			    'listening for kang and metric requests');
 		});
 
 		process.on('SIGINT', function () {
 			if (++nsigs == 1) {
-				if (kangserver)
-					kangserver.close();
+				monitor_server.close();
 				sock.close();
 				fastserver.close();
 			}
diff --git a/lib/fast_server.js b/lib/fast_server.js
index b4a0f21..e14a908 100644
--- a/lib/fast_server.js
+++ b/lib/fast_server.js
@@ -169,9 +169,11 @@ function FastServer(args)
 	mod_assertplus.object(args, 'args');
 	mod_assertplus.object(args.log, 'args.log');
 	mod_assertplus.object(args.server, 'args.server');
+	mod_assertplus.optionalObject(args.collector, 'args.collector');
 
 	this.fs_log = args.log;		/* logger */
 	this.fs_server = args.server;	/* server socket */
+	this.fs_collector = args.collector;		/* metric collector */
 	this.fs_handlers = {};		/* registered handlers, by name */
 	this.fs_conns = {};		/* active connections */
 	this.fs_msghandler = new FastMessageHandler({
@@ -201,6 +203,17 @@ function FastServer(args)
 	this.fs_nrequests_completed = 0;	/* count of reqs completed */
 	this.fs_nrequests_failed = 0;		/* count of reqs failed */
 
+	if (this.fs_collector) {
+		this.fs_collector.counter({
+			name: 'tcp_requests_completed',
+			help: 'count of tcp requests completed'
+		});
+		this.fs_collector.histogram({
+			name: 'tcp_request_time_ms',
+			help: 'total time to process tcp requests'
+		});
+	}
+
 	if (fastServerProvider === null) {
 		fastServerProvider = fastServerProviderInit();
 	}
@@ -840,6 +853,18 @@ FastServer.prototype.requestCleanup = function (request)
 
 	conn.fc_ncompleted++;
 	this.fs_nrequests_completed++;
+	if (this.fs_collector) {
+		/* Record metrics */
+		var histo, count;
+		var collector = this.fs_collector;
+		var latency = new Date() - request.fsr_tstarted;
+
+		count = collector.getCollector('tcp_requests_completed');
+		histo = collector.getCollector('tcp_request_time_ms');
+
+		count.increment();
+		histo.observe(latency);
+	}
 
 	if (request.fsr_handler !== null) {
 		request.fsr_handler.fh_ncompleted++;
diff --git a/package.json b/package.json
index 2cc02e8..657ebc5 100644
--- a/package.json
+++ b/package.json
@@ -1,13 +1,14 @@
 {
 	"name": "fast",
 	"description": "streaming JSON RPC over TCP",
-	"version": "2.2.4",
+	"version": "2.3.4",
 	"main": "./lib/fast.js",
 	"repository": {
 		"type": "git",
 		"url": "git://github.com/joyent/node-fast.git"
 	},
 	"dependencies": {
+		"artedi": "1.0.0",
 		"assert-plus": "1.0.0",
 		"bunyan": "^1.7.1",
 		"cmdutil": "^1.0.0",
@@ -18,6 +19,7 @@
 		"lstream": "0.0.4",
 		"microtime": "2.0.0",
 		"posix-getopt": "1.2.0",
+		"restify": "5.2.0",
 		"strsplit": "^1.0.0",
 		"vasync": "^1.6.3",
 		"verror": "^1.7.0"
diff --git a/test/tst.server.js b/test/tst.server.js
index ed8ee22..d80e6a6 100644
--- a/test/tst.server.js
+++ b/test/tst.server.js
@@ -13,6 +13,7 @@
  */
 
 var mod_assertplus = require('assert-plus');
+var mod_artedi = require('artedi');
 var mod_bunyan = require('bunyan');
 var mod_net = require('net');
 var mod_path = require('path');
@@ -57,6 +58,7 @@ function ServerTestContext()
 	this.ts_server = null;	/* fast server object */
 	this.ts_clients = [];	/* array of clients, each having properties */
 				/* "tsc_socket" and "tsc_client" */
+	this.ts_collector = null;	/* artedi metric collector */
 }
 
 ServerTestContext.prototype.connectClient = function (callback)
@@ -160,6 +162,9 @@ function runTestCase(testcase, callback)
 
 	tctx = new ServerTestContext();
 	tctx.ts_log = testLog.child({ 'testcase': testcase['name'] });
+	tctx.ts_collector = mod_artedi.createCollector({
+	    'labels': { 'component': 'FastServer' }
+	});
 	tctx.ts_socket = mod_net.createServer({ 'allowHalfOpen': true });
 	tctx.ts_server = new mod_fast.FastServer({
 	    'log': tctx.ts_log.child({ 'component': 'FastServer' }),
